### PR TITLE
magento/magento2#22870: ProductRepository fails to update an existing product with a changed SKU.

### PR DESCRIPTION
**Pull request by p-bystritsky**
### Fixed Issues (if relevant)

1. magento/magento2#22870: ProductRepository fails to update an existing product with a changed SKU.

### Manual testing scenarios (*)

1. See issue description.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: dev/tests/integration/testsuite/Magento/Catalog/Model/ProductRepositoryTest.php [hunk 1]
```diff
@@ -138,4 +141,28 @@ public function testSaveProductWithGalleryImage(): void
...
+    {
+        $newSku = 'simple-edited';
+        $productId = Bootstrap::getObjectManager()->get(ProductResource::class)->getIdBySku('simple');
+        $initialProduct = Bootstrap::getObjectManager()->create(Product::class)->load($productId);
+
+        $initialProduct->setSku($newSku);
+        $this->productRepository->save($initialProduct);
+
+        $updatedProduct = Bootstrap::getObjectManager()->create(Product::class);
+        $updatedProduct->load($productId);
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Catalog/Model/ProductRepositoryTest.php]: Please, avoid using deprecated methods. 
For loading a model, use a resource model and product model factory. So in result we have something like:

```suggestion
    // ProductResource is alias for \Magento\Catalog\Model\ResourceModel\Product
    // ProductFactory is alias for \Magento\Catalog\Model\ProductFactory
    $productFactory =  Bootstrap::getObjectManager()->get(ProductFactory::class)
    $updatedProduct = $productFactory->create()
    $productResource = Bootstrap::getObjectManager()->get(ProductResource::class)
    $productResource->load($updatedProduct, $productId)
```

**Review by rogyar — CHANGES_REQUESTED** [review]

**Comment by p-bystritsky** [comment]
@rogyar updated.

File: dev/tests/integration/testsuite/Magento/Catalog/Model/ProductRepositoryTest.php (hunk 1 above)
Comment (p-bystritsky) [reply dev/tests/integration/testsuite/Magento/Catalog/Model/ProductRepositoryTest.php]: @rogyar updated.

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 2]
```diff
@@ -570,7 +573,7 @@ public function save(ProductInterface $product, $saveOptions = false)
                 $product->getCategoryIds()
             );
         }
-        $this->removeProductFromLocalCache($product->getSku());
+        $this->removeProductFromLocalCacheBySku($product->getSku());
         unset($this->instancesById[$product->getId()]);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/ProductRepository.php]:
```suggestion
        $this->removeProductFromLocalCacheById($product->getId());
```

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 3]
```diff
@@ -584,7 +587,7 @@ public function delete(ProductInterface $product)
         $sku = $product->getSku();
         $productId = $product->getId();
         try {
-            $this->removeProductFromLocalCache($product->getSku());
+            $this->removeProductFromLocalCacheBySku($product->getSku());
             unset($this->instancesById[$product->getId()]);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/ProductRepository.php]:
```suggestion
        $this->removeProductFromLocalCacheById($product->getId());
```

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 4]
```diff
@@ -800,7 +814,7 @@ private function prepareSku(string $sku): string
     private function saveProduct($product): void
     {
         try {
-            $this->removeProductFromLocalCache($product->getSku());
+            $this->removeProductFromLocalCacheBySku($product->getSku());
             unset($this->instancesById[$product->getId()]);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/ProductRepository.php]:
```suggestion
        $this->removeProductFromLocalCacheById($product->getId());
```

File: app/code/Magento/Catalog/Model/ProductRepository.php [hunk 5]
```diff
@@ -595,7 +598,7 @@ public function delete(ProductInterface $product)
                 $e
             );
         }
-        $this->removeProductFromLocalCache($sku);
+        $this->removeProductFromLocalCacheBySku($sku);
         unset($this->instancesById[$productId]);
```
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/ProductRepository.php]:
```suggestion
        $this->removeProductFromLocalCacheById($productId);
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @p-bystritsky,
In general changes looks good. Could you review my suggestions? We don't need to have code duplication

**Comment by p-bystritsky** [comment]
> Hi @p-bystritsky,
> In general changes looks good. Could you review my suggestions? We don't need to have code duplication

@ihor-sviziev, thank you. Your suggestions committed.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @p-bystritsky,
Could you check failing tests? Looks like your change broke some functionality 
PS: very strange that we hadn’t these fails before fixing static tests, it it might be just coincidence

**Comment by p-bystritsky** [comment]
@ihor-sviziev fixed.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-5134](https://jira.corp.magento.com/browse/ENGCOM-5134) has been created to process this Pull Request

**Comment by soleksii** [comment]
:heavy_check_mark: QA Passed

**Before:**

before

**After:**

after

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-5134](https://jira.corp.magento.com/browse/ENGCOM-5134) has been created to process this Pull Request

**Pull request merged**